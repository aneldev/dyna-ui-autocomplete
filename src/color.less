.dyna-autocomplete {
	&.dyna-ui-field-wrapper-color- {

		&-color- {
			&WHITE_BLACK {
			}
			&WHITE_RED {
			}
			&BLACK_WHITE {
			}
			&TRANSPARENT_WHITE {
			}
			&ORANGE_WHITE {
			}
		}
	}
}
