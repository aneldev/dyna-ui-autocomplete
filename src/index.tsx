import {
  DynaAutoComplete, IDynaAutoCompleteProps,
  IAutoCompleteValue,
  EMode, EStyle, EColor,
} from "./DynaAutoComplete";

import {
  debounce
} from "./utils";

export {
  DynaAutoComplete, IDynaAutoCompleteProps,
  IAutoCompleteValue,
  EMode, EStyle, EColor,
  debounce,
}
