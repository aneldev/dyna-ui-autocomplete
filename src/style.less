.dyna-autocomplete {
	input {
		width: 100%;
		border-color: transparent;
		font-weight: bold;
		font-size: 16px;

		&:focus {
			outline: none;
		}
	}

	.dyna-autocomplete-menu {
		position: absolute;
		right: 0;
		left: 0;
		z-index: 1000;
		cursor: pointer;
	}

	&.dyna-ui-field-wrapper {

		&--mode-VIEW {
			// bug fix: for util.js:9 Uncaught TypeError: Cannot read property 'ownerDocument' of null
			// more info for this bug fix at https://github.com/reactjs/react-autocomplete/issues/261
			pointer-events: none;
		}

		&--style- {
			&INLINE_ROUNDED {

				.dyna-ui-picker-container {
					> * {
						&:first-child {
							border-top-left-radius: 8px;
							border-top-right-radius: 8px;
						}
						&:last-child {
							border-bottom-left-radius: 8px;
							border-bottom-right-radius: 8px;
						}
					}

				}

			}
		}

	}

}
