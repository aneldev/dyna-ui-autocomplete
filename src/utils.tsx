import * as React from 'react';
import Timer = NodeJS.Timer;

export const faIcon = (awesomeFontIconName:string, className:string='') => <i className={`fa fa-${awesomeFontIconName} ${className}`.trim()} aria-hidden="true" />;

export const debounce = (cbFunction: Function, timeout: number): Function => {

  let setTimerHolder: Timer = null;
  let lastCalled: Date = (0 as any);

  const runIt = () => {
    clearTimeout(setTimerHolder);
    setTimerHolder = null;
    lastCalled = new Date();
  };

  return () => {

    if (setTimerHolder) {
      clearTimeout(setTimerHolder);
    }

    if (Number(new Date()) - Number(lastCalled) > timeout) {
      runIt();
    } else {
      setTimerHolder = setTimeout(runIt, timeout);
    }

  }

};
