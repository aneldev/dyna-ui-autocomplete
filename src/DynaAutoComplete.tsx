import * as React from "react";
import * as ReactAutoComplete from 'react-autocomplete';
import {DynaFieldWrapper, EColor, EMode, EStyle, ESize} from "dyna-ui-field-wrapper"
import {DynaPickerContainer, EStyle as EPickerStyle, EColor as EPickerColor} from "dyna-ui-picker-container"
import {faIcon} from "./utils";

import "./style.less";

// help: https://github.com/reactjs/react-autocomplete

export {EMode, EColor, EStyle, ESize}

export interface IDynaAutoCompleteProps<TItem> {
  className?: string;
  name: string;
  style?: EStyle;
  color?: EColor;
  size?: ESize;
  mode?: EMode;
  label?: TContent;
  required?: TContent;
  isLoading?: boolean;
  isLoadingIcon?: TContent;
  items: TItem[];
  value: string;
  selectOnBlur?: boolean;
  inputProps?: React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>;
  getItemValue: (item: TItem) => string;
  renderItem: (item: TItem, isFocused: boolean) => JSX.Element;
  dropDownFilter?: (item: TItem, enteredText: string) => boolean;
  validationMessage?: TContent;
  footer?: TContent;
  onChange: (name: string, value: IAutoCompleteValue<TItem>) => void;
}

export type TContent = string | JSX.Element;

export interface IAutoCompleteValue<TItem> {
  value: string;
  item: TItem;
}

export class DynaAutoComplete<TItem> extends React.Component<IDynaAutoCompleteProps<TItem>> {
  static defaultProps: IDynaAutoCompleteProps<any> = {
    className: '',
    name: '',
    mode: EMode.EDIT,
    style: EStyle.INLINE_ROUNDED,
    color: EColor.WHITE_BLACK,
    size: ESize.MEDIUM,
    label: null,
    isLoading: false,
    isLoadingIcon: faIcon('circle-o-notch fa-spin'),
    items: [],
    value: "",
    inputProps: {},
    selectOnBlur: false,
    getItemValue: (item: any) => "",
    renderItem: (item: any, isFocused: boolean) => null,
    dropDownFilter: (item: any, enteredText: string) => true,
    validationMessage: null,
    footer: null,
    onChange: (name: string, value: IAutoCompleteValue<any>) => undefined,
  };

  private handlerOnChange(event: Event, value: string): void {
    if (this.props.mode === EMode.VIEW) return;
    const {name, items, onChange, getItemValue} = this.props;
    onChange(name, {
      value,
      item: items.find((item: TItem) => value === getItemValue(item)),
    });
  }

  private handlerOnSelect(value: string, item: TItem): void {
    if (this.props.mode === EMode.VIEW) return;
    if (this.props.value === value) return; // exit, nothing is changed
    const {name, onChange} = this.props;
    onChange(name, {
      value,
      item,
    });
  }

  private renderMenu(children: any): JSX.Element {
    if (this.props.mode === EMode.VIEW) return <div/>;
    if (!children || !children.length) return <div/>;

    return (
      <div className="dyna-autocomplete-menu">
        <DynaPickerContainer
          show
          style={EPickerStyle.ROUNDED}
          color={EPickerColor.WHITE_BLACK}
          responsive={false}
        >{children}</DynaPickerContainer>
      </div>
    )
  }

  public render(): JSX.Element {
    const {
      className,
      mode, style, color, size,
      label, required, isLoading, isLoadingIcon,
      items, value,
      selectOnBlur,
      inputProps,
      getItemValue, renderItem, dropDownFilter,
      validationMessage, footer,
    } = this.props;

    return (
      <DynaFieldWrapper
        className={`dyna-autocomplete ${className}`.trim()}
        style={style}
        color={color}
        size={size}
        mode={mode}
        inputElementSelector="input"
        label={label}
        isLoading={isLoading ? isLoadingIcon : null}
        required={required}
        validationMessage={validationMessage}
        footer={footer}
      >
        <ReactAutoComplete
          items={items}
          value={value}
          enabled={mode === EMode.EDIT}
          selectOnBlur={selectOnBlur}
          getItemValue={getItemValue}
          renderMenu={this.renderMenu.bind(this)}
          renderItem={renderItem}
          shouldItemRender={dropDownFilter}
          inputProps={inputProps}
          onChange={this.handlerOnChange.bind(this)}
          onSelect={this.handlerOnSelect.bind(this)}
        />
      </DynaFieldWrapper>
    );
  }
}
