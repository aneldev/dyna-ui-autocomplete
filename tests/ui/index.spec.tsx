import { configure } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

import * as React from 'react';
import * as enzyme from 'enzyme';

import {DynaAutoComplete} from '../../src';
import {IAutoCompleteValue} from "../../src/DynaAutoComplete";

describe('Home', () => {
  let wrapper;

  it('has expected content with deep render', () => {
    wrapper = enzyme.shallow(
      (
        <DynaAutoComplete
          name="countries"
          items={[]}
          value="Austria"
          getItemValue={(item: any) => ""}
          renderItem={(item: any, isFocused: boolean) => <span>item.label</span>}
          onChange={(name: string, value: IAutoCompleteValue<any>) => undefined}
        />
      ),
      {}
    );

    expect(wrapper).toMatchSnapshot();
  });
});
