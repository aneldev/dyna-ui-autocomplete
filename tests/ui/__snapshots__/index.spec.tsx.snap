// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home has expected content with deep render 1`] = `
<DynaFieldWrapper
  className="dyna-autocomplete"
  color="WHITE_BLACK"
  footer={null}
  inputElementSelector="input"
  isLoading={null}
  label={null}
  required={null}
  style="INLINE_ROUNDED"
  validationMessage={null}
>
  <Autocomplete
    autoHighlight={true}
    getItemValue={[Function]}
    inputProps={
      Object {
        "onBlur": [Function],
      }
    }
    items={Array []}
    menuStyle={
      Object {
        "background": "rgba(255, 255, 255, 0.9)",
        "borderRadius": "3px",
        "boxShadow": "0 2px 12px rgba(0, 0, 0, 0.1)",
        "fontSize": "90%",
        "maxHeight": "50%",
        "overflow": "auto",
        "padding": "2px 0",
        "position": "fixed",
      }
    }
    onChange={[Function]}
    onMenuVisibilityChange={[Function]}
    onSelect={[Function]}
    renderInput={[Function]}
    renderItem={[Function]}
    renderMenu={[Function]}
    selectOnBlur={false}
    shouldItemRender={[Function]}
    value="Austria"
    wrapperProps={Object {}}
    wrapperStyle={
      Object {
        "display": "inline-block",
      }
    }
  />
</DynaFieldWrapper>
`;
